README
lightum - MacBook automatic light sensor daemon
(c)2011-2012 Pau Oliva Fora - pof[at]eslack(.)org

lightum is a daemon to control the keyboard brightness and monitor backlight on
MacBook based laptops.

By default it will run in "auto" mode, this means it will read data from the
ambient light sensor (located on the iSight camera) and automatically:

	a) light up or dim the keyboard brightness 
	b) light up or dim the the video backlight 

It will also dim the keyboard brightness and screen backlight when the computer
is not used for a configurable amount of seconds (5 by default).

If you choose to run it in "manual" mode, it will not query the ambient light
sensor, but instead it will use the brightness and backlight values you set
manually using the function keys.

When first launched lightum will create a configuration file in the folder
~/.config/lightum/lightum.conf, inside the user's home directory with the
default configuration values. When launched without parameters, lightum will
read the configuration from this config file, but the configuration values can
be overwritten via the command line:

Usage:  lightum [OPTION]...
     -m 4..255 : maximum brightness value in auto mode (default=255)
     -n 0..3   : minimum brightness value in auto mode (default=0)
     -M 7..15  : maximum backlight value in auto mode (default=15)
     -N 1..6   : minimum backlight value in auto mode (default=1)
     -p num    : number of milliseconds between light sensor polls (default=300)
     -i num    : power off keyboard light on session idle seconds (0 to disable)
     -I num    : turn down screen backlight on session idle seconds (0 to disable)
     -w num    : 1 manage brightness, 2 manage backlight, 3 both (default:3)
     -x        : manual mode (will honor the brightness value set with Fn keys)
     -l        : fully dim the screen backlight when session is idle
     -u        : do not ignore brightness changes happening outside lightum
     -U        : ignore session information from ConsoleKit
     -s        : power off keyboard light when screen saver is active
     -f        : run in foreground (do not daemonize)
     -v        : verbose mode, useful for debugging with -f and -d
     -d num    : debug mode: 1 brightness, 2 backlight, 3 both


Official Ubuntu packages are available in poliva/lightum-mba ppa:

     sudo add-apt-repository ppa:poliva/lightum-mba
     sudo apt-get update
     sudo apt-get install lightum

Enjoy! :)


debian/changelog
lightum (2.3.1-ubuntu1) precise; urgency=low

  * updated to version 2.3.1 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 11 Jun 2012 18:20:58 +0200

lightum (2.3-ubuntu1) precise; urgency=low

  * updated to version 2.3 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 14 May 2012 21:11:21 +0200

lightum (2.2.1-ubuntu1) precise; urgency=low

  * updated to version 2.2.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 13 May 2012 17:35:57 +0200

lightum (2.2-ubuntu1) precise; urgency=low

  * updated to version 2.2
  * added libc6-dev build time dependancy for math.h

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 13 May 2012 16:50:49 +0200

lightum (2.1.1-ubuntu1) precise; urgency=low

  * updated to version 2.1.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 11 May 2012 20:11:02 +0200

lightum (2.1-ubuntu1) precise; urgency=low

  * updated to version 2.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 09 May 2012 21:08:30 +0200

lightum (2.0.4-ubuntu1) precise; urgency=low

  * updated to version 2.0.4 (fixes a memory leak)

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 04 May 2012 22:05:20 +0200

lightum (2.0.3-ubuntu1) precise; urgency=low

  * updated to version 2.0.3
  * added debian/watch file

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 03 May 2012 16:27:42 +0200

lightum (2.0.2-ubuntu1) precise; urgency=low

  * updated to version 2.0.2
  * added libdbus-glib-1-dev as build time dependancy
  * added consolekit as run time dependancy
  * updated program description

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 02 May 2012 17:46:59 +0200

lightum (2.0.1-ubuntu1) precise; urgency=low

  * updated to version 2.0.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 01 May 2012 11:49:30 +0200

lightum (2.0-ubuntu1) precise; urgency=low

  * updated to version 2.0

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 01 May 2012 01:16:54 +0200

lightum (1.7-ubuntu1) precise; urgency=low

  * updated to version 1.7

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 26 Apr 2012 22:28:07 +0200

lightum (1.6-ubuntu1) oneiric; urgency=low

  * updated to version 1.6

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Thu, 26 Apr 2012 16:52:54 +0200

lightum (1.5-ubuntu1) oneiric; urgency=low

  * updated to version 1.5

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 13 Dec 2011 15:45:33 +0100

lightum (1.4.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.4.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Mon, 12 Dec 2011 00:35:01 +0100

lightum (1.4-ubuntu1) oneiric; urgency=low

  * updated to version 1.4

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sat, 10 Dec 2011 03:13:28 +0100

lightum (1.3.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.3.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 09 Dec 2011 23:35:02 +0100

lightum (1.3-ubuntu1) oneiric; urgency=low

  * updated to version 1.3

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Fri, 09 Dec 2011 21:00:39 +0100

lightum (1.2-ubuntu1) oneiric; urgency=low

  * updated to version 1.2, removed libdbus-glib-1-dev dependancy, added libxss-dev dependancy.

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Wed, 07 Dec 2011 03:27:00 +0100

lightum (1.1-ubuntu4) oneiric; urgency=low

  * fix build time dependancy, hopefully. Package name is libdbus-glib-1-dev, not dbus-glib-1.

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 05:12:17 +0100

lightum (1.1-ubuntu3) oneiric; urgency=low

  * fix build time dependancy, again: dbus-glib-1 is included in Makefile pkgconfig, but not really needed. 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:57:04 +0100

lightum (1.1-ubuntu2) oneiric; urgency=low

  * fix build time dependancy on debian control file (dbus/dbus.h) 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:38:48 +0100

lightum (1.1-ubuntu1) oneiric; urgency=low

  * updated to version 1.1

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Tue, 06 Dec 2011 04:17:53 +0100

lightum (1.0-ubuntu2) oneiric; urgency=low

  * Fix Makefile for launchpad build 

 -- Pau Oliva Fora (pof) <pau@eslack.org>  Sun, 04 Dec 2011 12:04:07 +0100

lightum (1.0-ubuntu1) oneiric; urgency=low

  * Initial release

 -- Pau Oliva Fora (pof) <pau@eslack.org> Sun, 04 Dec 2011 05:23:38 +0100 


dbus.c
/*
 *   MacBook automatic light sensor daemon
 *   Copyright 2011 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <dbus/dbus.h>

int get_screensaver_active() {

	DBusConnection *connection;
	DBusError error;
	DBusMessage *message;
	DBusMessageIter iter;
	DBusBusType type;
	DBusMessage *reply;

	int retval=0;
	const char *name = NULL;
	char arg[]="org.gnome.ScreenSaver.GetActive";
	name=arg;
	const char *dest = "org.gnome.ScreenSaver";
	const char *path = "/org/gnome/ScreenSaver";

	type = DBUS_BUS_SESSION;

	int reply_timeout;
	reply_timeout = 2000;

	dbus_error_init (&error);
	connection = dbus_bus_get (type, &error);

	if (connection == NULL) {
		fprintf (stderr, "Failed to open connection to message bus: %s\n", error.message);
		dbus_error_free (&error);
		return (-1);
	}

	char *last_dot;
	last_dot = strrchr (name, '.');
	*last_dot = '\0';

	message = dbus_message_new_method_call (NULL,path,name,last_dot + 1);
	dbus_message_set_auto_start (message, TRUE);

	if (message == NULL)
	{
		fprintf (stderr, "Couldn't allocate D-Bus message\n");
		return (-1);
	}

	if (dest && !dbus_message_set_destination (message, dest))
	{
		fprintf (stderr, "Not enough memory\n");
		return (-1);
	}

	dbus_message_iter_init_append (message, &iter);

	dbus_error_init (&error);
	reply = dbus_connection_send_with_reply_and_block (connection, message, reply_timeout, &error);
	if (dbus_error_is_set (&error)) {
		fprintf (stderr, "Error %s: %s\n",error.name,error.message);
		return (-1);
	}

	if (reply)
	{
		dbus_message_iter_init (reply, &iter);
		dbus_bool_t val;
		dbus_message_iter_get_basic (&iter, &val);
		retval = val;
		dbus_message_unref (reply);
	}

	dbus_message_unref (message);
	dbus_connection_unref (connection);

	return retval;
}


int set_keyboard_brightness_value(int brightness) {

	DBusConnection *connection;
	DBusError error;
	DBusMessage *message;
	DBusMessageIter iter;
	DBusBusType type;

	const char *name = NULL;
	char arg[]="org.freedesktop.UPower.KbdBacklight.SetBrightness";
	name=arg;
	const char *dest="org.freedesktop.UPower";
	const char *path="/org/freedesktop/UPower/KbdBacklight";

	type = DBUS_BUS_SYSTEM;

	dbus_error_init (&error);
	connection = dbus_bus_get (type, &error);

	if (connection == NULL) {
		fprintf (stderr, "Failed to open connection to message bus: %s\n", error.message);
		dbus_error_free (&error);
		return (-1);
	}

	char *last_dot;
	last_dot = strrchr (name, '.');
	*last_dot = '\0';

	message = dbus_message_new_method_call (NULL,path,name,last_dot + 1);
	dbus_message_set_auto_start (message, TRUE);

	if (message == NULL)
	{
		fprintf (stderr, "Couldn't allocate D-Bus message\n");
		return (-1);
	}

	if (dest && !dbus_message_set_destination (message, dest))
	{
		fprintf (stderr, "Not enough memory\n");
		return (-1);
	}

	dbus_message_iter_init_append (message, &iter);

	dbus_int32_t int32;
	int32 = brightness;
	dbus_message_iter_append_basic (&iter, DBUS_TYPE_INT32, &int32);

	dbus_connection_send (connection, message, NULL);
	dbus_connection_flush (connection);
	dbus_message_unref (message);
	dbus_connection_unref (connection);
	return 0;
}


int dbus_get_screen_backlight_value() {

	DBusConnection *connection;
	DBusError error;
	DBusMessage *message;
	DBusMessageIter iter;
	DBusBusType type;
	DBusMessage *reply;

	int retval=0;
	const char *name = NULL;
	char arg[]="org.gnome.SettingsDaemon.Power.Screen.GetPercentage";
	name=arg;
	const char *dest = "org.gnome.SettingsDaemon";
	const char *path = "/org/gnome/SettingsDaemon/Power";

	type = DBUS_BUS_SESSION;

	int reply_timeout;
	reply_timeout = 2000;

	dbus_error_init (&error);
	connection = dbus_bus_get (type, &error);

	if (connection == NULL) {
		fprintf (stderr, "Failed to open connection to message bus: %s\n", error.message);
		dbus_error_free (&error);
		return (-1);
	}

	char *last_dot;
	last_dot = strrchr (name, '.');
	*last_dot = '\0';

	message = dbus_message_new_method_call (NULL,path,name,last_dot + 1);
	dbus_message_set_auto_start (message, TRUE);

	if (message == NULL)
	{
		fprintf (stderr, "Couldn't allocate D-Bus message\n");
		return (-1);
	}

	if (dest && !dbus_message_set_destination (message, dest))
	{
		fprintf (stderr, "Not enough memory\n");
		return (-1);
	}

	dbus_message_iter_init_append (message, &iter);

	dbus_error_init (&error);
	reply = dbus_connection_send_with_reply_and_block (connection, message, reply_timeout, &error);
	if (dbus_error_is_set (&error)) {
		fprintf (stderr, "Error %s: %s\n",error.name,error.message);
		return (-1);
	}

	if (reply)
	{
		dbus_message_iter_init (reply, &iter);
		dbus_bool_t val;
		dbus_message_iter_get_basic (&iter, &val);
		retval = val;
		dbus_message_unref (reply);
	}

	dbus_message_unref (message);
	dbus_connection_unref (connection);

	return retval;
}

int dbus_set_screen_backlight_value_gnome(int backlight) {

	DBusConnection *connection;
	DBusError error;
	DBusMessage *message;
	DBusMessageIter iter;
	DBusBusType type;
	DBusMessage *reply;

	int retval=0;
	const char *name = NULL;
	char arg[]="org.gnome.SettingsDaemon.Power.Screen.SetPercentage";
	name=arg;
	const char *dest = "org.gnome.SettingsDaemon";
	const char *path = "/org/gnome/SettingsDaemon/Power";

	type = DBUS_BUS_SESSION;

	int reply_timeout;
	reply_timeout = 2000;

	dbus_error_init (&error);
	connection = dbus_bus_get (type, &error);

	if (connection == NULL) {
		fprintf (stderr, "Failed to open connection to message bus: %s\n", error.message);
		dbus_error_free (&error);
		return (-1);
	}

	char *last_dot;
	last_dot = strrchr (name, '.');
	*last_dot = '\0';

	message = dbus_message_new_method_call (NULL,path,name,last_dot + 1);
	dbus_message_set_auto_start (message, TRUE);

	if (message == NULL)
	{
		fprintf (stderr, "Couldn't allocate D-Bus message\n");
		return (-1);
	}

	if (dest && !dbus_message_set_destination (message, dest))
	{
		fprintf (stderr, "Not enough memory\n");
		return (-1);
	}


	dbus_message_iter_init_append (message, &iter);

	dbus_uint32_t uint32;
	uint32 = backlight;
	dbus_message_iter_append_basic (&iter, DBUS_TYPE_UINT32, &uint32);

	dbus_error_init (&error);
	reply = dbus_connection_send_with_reply_and_block (connection, message, reply_timeout, &error);
	if (dbus_error_is_set (&error)) {
		fprintf (stderr, "Error %s: %s\n",error.name,error.message);
		return (-1);
	}

	if (reply)
	{
		dbus_message_iter_init (reply, &iter);
		dbus_bool_t val;
		dbus_message_iter_get_basic (&iter, &val);
		retval = val;
		dbus_message_unref (reply);
	}

	dbus_message_unref (message);
	dbus_connection_unref (connection);

	return retval;
}

int dbus_set_screen_backlight_value_kde(int backlight) {

	DBusConnection *connection;
	DBusError error;
	DBusMessage *message;
	DBusMessageIter iter;
	DBusBusType type;
	DBusMessage *reply;

	int retval=0;
	const char *name = NULL;
        char arg[]="org.kde.Solid.PowerManagement.setBrightness";
        name=arg;
        const char *dest = "org.kde.Solid.PowerManagement";
        const char *path = "/org/kde/Solid/PowerManagement"; 

	type = DBUS_BUS_SESSION;

	int reply_timeout;
	reply_timeout = 2000;

	dbus_error_init (&error);
	connection = dbus_bus_get (type, &error);

	if (connection == NULL) {
		fprintf (stderr, "Failed to open connection to message bus: %s\n", error.message);
		dbus_error_free (&error);
		return (-1);
	}

	char *last_dot;
	last_dot = strrchr (name, '.');
	*last_dot = '\0';

	message = dbus_message_new_method_call (NULL,path,name,last_dot + 1);
	dbus_message_set_auto_start (message, TRUE);

	if (message == NULL)
	{
		fprintf (stderr, "Couldn't allocate D-Bus message\n");
		return (-1);
	}

	if (dest && !dbus_message_set_destination (message, dest))
	{
		fprintf (stderr, "Not enough memory\n");
		return (-1);
	}


	dbus_message_iter_init_append (message, &iter);

	dbus_int32_t int32;
	int32 = backlight;
	dbus_message_iter_append_basic (&iter, DBUS_TYPE_INT32, &int32);

	dbus_error_init (&error);
	reply = dbus_connection_send_with_reply_and_block (connection, message, reply_timeout, &error);
	if (dbus_error_is_set (&error)) {
		fprintf (stderr, "Error %s: %s\n",error.name,error.message);
		return (-1);
	}

	if (reply)
	{
		dbus_message_iter_init (reply, &iter);
		//dbus_bool_t val;
		//dbus_message_iter_get_basic (&iter, &val);
		//retval = val;
		dbus_message_unref (reply);
	}

	dbus_message_unref (message);
	dbus_connection_unref (connection);

	return retval;
}

int dbus_set_screen_backlight_value(int backlight, int backend) {

	int ret=-1;

	if (backend == 0) ret = dbus_set_screen_backlight_value_gnome(backlight); 
	if (backend == 1) ret = dbus_set_screen_backlight_value_kde(backlight);

	return ret;
}


lightum.h
#include <X11/Xlib.h>
#include <dbus/dbus-glib-lowlevel.h>

#define CONFIG_PATH ".config/lightum/"
#define CONFIG_FILE "lightum.conf"
#define PID_FILE "lightum.pid"

#ifndef FALSE
#define FALSE   (0)
#endif

#ifndef TRUE
#define TRUE    (!FALSE)
#endif

/* config file */
typedef struct {
	int manualmode;
	int queryscreensaver;
	int maxbrightness;
	int minbrightness;
	int polltime;
	int idleoff;
	int ignoreuser;
	int minbacklight;
	int maxbacklight;
	int screenidle;
	int workmode;
	int ignoresession;
	int fulldim;
} conf_data;

/* configfile.c */
conf_data config_parse();
int file_exists(char* file);
char* default_config_dir();
int create_config_dir(char* path);

/* functions.c */
int get_keyboard_brightness_value();
int get_light_sensor_value();
int get_screen_backlight_value();
int dbus_to_acpi_backlight(int backlight);
int acpi_to_dbus_backlight(int backlight);
int set_screen_backlight_value(int backlight, int backend);
int calculate_keyboard_brightness_value(int light, int maxlight, int minlight);
int calculate_screen_backlight_value(int light, int maxbacklight, int minbacklight);
void fading(int from, int to);
void backlight_fading(int from, int to, int backend);
float get_session_idle_time(Display *display);
void signal_installer();
int remove_pid_file();
int create_pid_file();

/* dbus.c */
int get_screensaver_active();
int set_keyboard_brightness_value(int brightness);
int dbus_get_screen_backlight_value();
int dbus_set_screen_backlight_value_gnome(int backlight);
int dbus_set_screen_backlight_value_kde(int backlight);
int dbus_set_screen_backlight_value(int backlight, int backend);

/* dbus-session.c */
DBusGConnection* get_dbus_connection();
DBusGProxy* get_dbus_proxy_manager(DBusGConnection *connection);
DBusGProxy* get_dbus_proxy_session(DBusGConnection *connection, DBusGProxy *proxy_manager);
int get_session_active (DBusGProxy *proxy_session);


functions.c
/*
 *   MacBook automatic light sensor daemon
 *   Copyright 2011 Pau Oliva Fora <pof@eslack.org>
 *
 *   This program is free software; you can redistribute it and/or
 *   modify it under the terms of the GNU General Public License
 *   as published by the Free Software Foundation; either version 2
 *   of the License, or (at your option) any later version.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <unistd.h>
#include <X11/Xlib.h>
#include <X11/extensions/scrnsaver.h>
#include <signal.h>
#include <sys/stat.h>
#include <string.h>
#include <math.h>

#include "lightum.h"

extern int reloadconfig;

int get_keyboard_brightness_value() {

	int fd;
	char buf[5];
	const char *kbd_backlight="/sys/devices/platform/applesmc.768/leds/smc::kbd_backlight/brightness";
	ssize_t cnt;

	/* read light sensor value */
	fd = open(kbd_backlight, O_RDONLY);
	if (fd < 0) {
		perror (kbd_backlight);
                fprintf (stderr, "Can't open %s\n",kbd_backlight);
		exit(1);
	}
	cnt=read(fd, buf, sizeof(buf)-1);
	buf[cnt]='\0';
	close(fd);

	return atoi(buf);
}

int get_light_sensor_value() {

	int fd;
	size_t i,n=0;
	char buf[10];
	char a_light[10];
	const char *light_sensor="/sys/devices/platform/applesmc.768/light";
	ssize_t cnt;

	/* read light sensor value */
	fd = open(light_sensor, O_RDONLY);
	if (fd < 0) {
		perror (light_sensor);
                fprintf (stderr, "Can't open %s\n",light_sensor);
		exit(1);
	}
	cnt=read(fd, buf, sizeof(buf)-1);
	buf[cnt]='\0';
	close(fd);

	/* convert light sensor string value to integer */
	for (i=0;buf[i]!='\0';i++) {
		if (buf[i]==',') break;
		if (buf[i]!='(') {
			a_light[n]=buf[i];
			n++;
		}
	}
	a_light[n]='\0';

	return atoi(a_light);
}

int get_screen_backlight_value() {

	int fd, actual_backlight, max_backlight;
	char buf[5];
	struct stat tmp;
	const char *scr_backlight="/sys/class/backlight/acpi_video0/actual_brightness";
	const char *scr_maxbacklight="/sys/class/backlight/acpi_video0/max_brightness";
	ssize_t cnt;
        int ret=0, backlight=15;

	if (stat(scr_backlight, &tmp) == 0) {

		/* read screen backlight value */
		fd = open(scr_backlight, O_RDONLY);
		if (fd < 0) {
			perror (scr_backlight);
			fprintf (stderr, "Can't open %s\n",scr_backlight);
			exit(1);
		}
		cnt=read(fd, buf, sizeof(buf)-1);
		buf[cnt]='\0';
		close(fd);
		actual_backlight=atoi(buf);

		/* read screen max backlight value */
		fd = open(scr_maxbacklight, O_RDONLY);
		if (fd < 0) {
			perror (scr_backlight);
			fprintf (stderr, "Can't open %s\n",scr_backlight);
			exit(1);
		}
		cnt=read(fd, buf, sizeof(buf)-1);
		buf[cnt]='\0';
		close(fd);
		max_backlight=atoi(buf);

		/* make sure we always retrn a value between 0 and 15 */
		return (15*actual_backlight)/max_backlight;

	} else {
		// fallback to read the current screen backlight value using dbus
		// we prefer to avoid this because it forks gsd-backlight-helper
		// from gnome-settings-daemon on each call.

		backlight = dbus_get_screen_backlight_value();
		ret = dbus_to_acpi_backlight(backlight);
		return ret;

	}
}

int dbus_to_acpi_backlight(int backlight) {

	int value=15;
	if (backlight < 6 ) value=0;
	else if (backlight < 13) value=1;
	else if (backlight < 20) value=2;
	else if (backlight < 26) value=3;
	else if (backlight < 33) value=4;
	else if (backlight < 40) value=5;
	else if (backlight < 46) value=6;
	else if (backlight < 53) value=7;
	else if (backlight < 60) value=8;
	else if (backlight < 66) value=9;
	else if (backlight < 73) value=10;
	else if (backlight < 80) value=11;
	else if (backlight < 86) value=12;
	else if (backlight < 93) value=13;
	else if (backlight < 100) value=14;
	else if (backlight == 100) value=15;
	return value;
}

int acpi_to_dbus_backlight(int backlight) {

	int value=100;
	if (backlight == 15) value=100;
	else if (backlight == 14) value=96;
	else if (backlight == 13) value=88;
	else if (backlight == 12) value=82;
	else if (backlight == 11) value=75;
	else if (backlight == 10) value=68;
	else if (backlight == 9) value=62;
	else if (backlight == 8) value=55;
	else if (backlight == 7) value=48;
	else if (backlight == 6) value=42;
	else if (backlight == 5) value=35;
	else if (backlight == 4) value=28;
	else if (backlight == 3) value=22;
	else if (backlight == 2) value=15;
	else if (backlight == 1) value=8;
	else if (backlight == 0) value=0;
	return value;
}

int set_screen_backlight_value(int backlight, int backend) {

	int value, ret;

	value = acpi_to_dbus_backlight(backlight);
	ret = dbus_set_screen_backlight_value(value, backend);

	return ret;
}

int calculate_keyboard_brightness_value(int light, int maxlight, int minlight) {

	int brightness=0;
	
	if (light == 0) brightness=maxlight;
	else brightness = (maxlight/2)/light;
	if (brightness < minlight) brightness=minlight;

	return brightness;
}

int calculate_screen_backlight_value(int light, int maxbacklight, int minbacklight) {

	int backlight=0;

        backlight=minbacklight+(log(light+1)/log(256))*(maxbacklight+1-minbacklight);
        if (backlight < minbacklight) backlight=minbacklight;
        if (backlight > maxbacklight) backlight=maxbacklight;

	return backlight;
}

void fading(int from, int to) {

	int step;

	if (from == -1)
		from = get_keyboard_brightness_value();

	if (from > to) {
		step=(from-to)/4;
		set_keyboard_brightness_value(from-step);
		usleep(100000);
		set_keyboard_brightness_value(from-step*2);
		usleep(100000);
		set_keyboard_brightness_value(from-step*3);
		usleep(100000);
		set_keyboard_brightness_value(to);
	}

	if (from < to) {
		step=(to-from)/4;
		set_keyboard_brightness_value(to-step*3);
		usleep(20000);
		set_keyboard_brightness_value(to-step*2);
		usleep(20000);
		set_keyboard_brightness_value(to-step);
		usleep(20000);
		set_keyboard_brightness_value(to);
	}
}

void backlight_fading(int from, int to, int backend) {

	int step;

	if (from == -1)
		from = get_screen_backlight_value();

	if (from > to) {
		step=(from-to)/4;
		set_screen_backlight_value(from-step, backend);
		usleep(100000);
		set_screen_backlight_value(from-step*2, backend);
		usleep(100000);
		set_screen_backlight_value(from-step*3, backend);
		usleep(100000);
		set_screen_backlight_value(to, backend);
	}

	if (from < to) {
		step=(to-from)/4;
		set_screen_backlight_value(to-step*3, backend);
		usleep(20000);
		set_screen_backlight_value(to-step*2, backend);
		usleep(20000);
		set_screen_backlight_value(to-step, backend);
		usleep(20000);
		set_screen_backlight_value(to, backend);
	}
}


float get_session_idle_time(Display *display) {

	XScreenSaverInfo info;
	float seconds;

	XScreenSaverQueryInfo(display, DefaultRootWindow(display), &info);
	seconds = (float)info.idle/1000.0f;
	return(seconds);
}

void signal_handler(int sig) {

	(void) sig;
	set_keyboard_brightness_value(0);
	remove_pid_file();
	printf("Killed!\n");
	exit(1);
}

void config_reload(int sig) {

	(void) sig;
	reloadconfig=1;
}

void signal_installer() {

	signal(SIGINT, signal_handler);
	signal(SIGTERM, signal_handler);
	signal(SIGHUP, signal_handler);
	signal(SIGQUIT, signal_handler);
	signal(SIGCHLD, signal_handler);
	signal(SIGABRT, signal_handler);
	signal(SIGUSR1, config_reload);
}

char* default_pid_file() {
	char* home = getenv("HOME");
	char* file = malloc(strlen(home) + strlen(CONFIG_PATH) + strlen(PID_FILE) + 2);
	strcpy(file, "");
	strcat(file, home);
	strcat(file, "/");
	strcat(file, CONFIG_PATH);
	strcat(file, PID_FILE);
	return file;
}

int create_pid_file() {

	int fd;
	char *path, *pidfile;
	char buf[100];
	ssize_t cnt;
	char* procpid = malloc( sizeof(buf) + 15 );

	path = default_config_dir();
	create_config_dir(path);

	pidfile = default_pid_file();

	if (file_exists(pidfile)) {

		// check if /proc/{pid}/cmdline exists and contains lightum
		// if it does, means lightum is already running, so we exit cowardly
		// if it does not contain lightum, then we remove the old pid file and continue

		fd = open(pidfile, O_RDONLY);
		if (fd < 0) {
			fprintf (stderr,"Could not open pid file: %s\n", pidfile);
			return FALSE;
		}
		cnt=read(fd, buf, sizeof(buf)-1);
		buf[cnt]='\0';
		
		close(fd);

		strcpy(procpid, "");
		strcat(procpid, "/proc/");
		strcat(procpid, buf);
		strcat(procpid, "/cmdline");

		if (file_exists(procpid)) {
			fd = open(procpid, O_RDONLY);
			if (fd < 0) {
				fprintf (stderr,"Could not open file: %s\n", procpid);
				return FALSE;
			}

			cnt=read(fd, buf, sizeof(buf)-1);
			buf[cnt]='\0';
			
			close(fd);

			if (strstr(buf,"lightum") != NULL) {
				fprintf (stderr,"Refusing to start as lightum is already running\n");
				return FALSE;
			} else {
				if (!remove_pid_file()) 
					return FALSE;
			}
		}
	}

	fd = open(pidfile, O_WRONLY | O_CREAT | O_TRUNC, 0644);
	if (fd < 0 ) {
		fprintf(stderr,"Could not write pid file: %s\n", pidfile);
		return FALSE;
	}

	sprintf( buf, "%d", getpid() );
	if (write(fd, buf, strlen(buf)) < 1) {
		perror("Something wrong happening while writing pid file");
		close(fd);
		return FALSE;
	}
	close(fd);

	free(procpid);

	return TRUE;
}

int remove_pid_file() {

	char *pidfile;

	pidfile = default_pid_file();

	if (!file_exists(pidfile)) {
		fprintf (stderr,"pid file does not exist: %s\n", pidfile);
		return TRUE;
	}

	if (unlink(pidfile) != 0) {
		fprintf (stderr,"Could not delete pid file: %s\n", pidfile);
		return FALSE;
	}
	return TRUE;
}
